### DISCUSSION TITLE
How to Work with Data Types

### DISCUSSION THREAD

**Discussion by adamgranthendry** [pyvista/pyvista#4921] [Q&A]
I am confused as how to extract values from `pyvista` data types. For example, consider the `examples.load_random_hills()` dataset. How do I get the `Elevations` at, say, the first, third, and fifth points in the dataset? It would be great if `pyvisa` could add examples in the docs of how to use the data types (I'm personally a little confused about how they work and when to access items via keys or via methods).

For example, `extract_points(0)` doesn't return the 0th point of the hills data, but rather an `UnstructuredGrid` with 2 cells, 4 points, and 4 arrays, which confuses me:

**Code:**
```python
import pyvista as pv
from pyvista import examples

data = examples.load_random_hills()
print(data.extract_points(0))
```

**Output:**
```
UnstructuredGrid (0x2c8a33d3340)
  N Cells:      2
  N Points:     4
  X Bounds:     0.000e+00, 2.020e-01
  Y Bounds:     2.980e+01, 3.000e+01
  Z Bounds:     1.028e+01, 1.035e+01
  N Arrays:     4
```

**Comment by MatthewFlamm** [comment]
https://docs.pyvista.org/core/filters.html?highlight=extract_points#pyvista.DataSetFilters.extract_points

This filter gives you the point and the cells that contain that point.

It sounds like you want` data['Elevation'][0]`.  The data on the mesh do not exist directly on the points themselves in this sense.  The data is stored as an array that is valid on the underlying mesh.  The filters allow you to extract different features or add new information, and sometimes this creates a new mesh and/or new data entirely.

The documentation has a discussion here https://docs.pyvista.org/user-guide/what-is-a-mesh.html#what-is-a-mesh, specifically in the "What are attributes?" section.  It sounds like there should be an expansion on this topic, as this section mostly focuses on setting the data and not accessing.

**Comment by adamgranthendry** [comment]
@MatthewFlamm Thank you for your help! I see: the docs state `extract_points` returns both cells and points. For context, if you see the example I wrote in [Project Surface Mesh to Cell/Finite Plane pyvista/pyvista-support#461](https://github.com/pyvista/pyvista-support/issues/461), I was having trouble at this section of the code:

```python
    elevations = np.hstack(
        [
            data.extract_points(ind, include_cells=False)["Elevation"]
            for ind in tqdm(data_inds)
        ]
    )

    # Mask points on planes
    planes.cell_arrays["Elevation"] = np.zeros((planes.n_cells,))
    planes.cell_arrays["Elevation"][pt_inds] = elevations
    planes.set_active_scalars("Elevation")  # Probably not necessary, but just in case
```

I used `include_cels=False` to ensure only points would be returned by `extract_points`, however my solution for mapping points to the plane here is rather slow and inelegant (in my opinion).

It would be great if we could expand upon accessing data in the docs as users need to access data regularly. I would like to learn more about how the data types are structured in `pyvista`.

@akaszynski What do you think?

**Comment by akaszynski** [comment]
Elevation data can be obtained from `point_arrays`.  Our documentation talks about how this is basically a "smart" dictionary that links the vtk data directly to numpy arrays.

Therefore, you can replace:

```py
    elevations = np.hstack(
        [
            data.extract_points(ind, include_cells=False)["Elevation"]
            for ind in tqdm(data_inds)
        ]
    )
```

With simply:

```py
elevations = data.point_arrays["Elevation"][data_inds]
```

Data obtained is identical.  Effectively, you want to rearrange the existing point elevation based on `data_inds` from `pykdtree`.  Since `data.point_arrays["Elevation"]` is a 1:1 mapping of elevation data to point index (it's the only way point data can be stored), then you can simply index it.

**Comment by akaszynski** [comment]
> It would be great if pyvisa could add examples in the docs of how to use the data types

Clearly we need to add better examples since this is confusing.  Feel free to open this up as an issue.  We're working on your earlier one.

**Comment by adamgranthendry** [comment]
@akaszynski and @MatthewFlamm Thank you very much! Yes, I will open an issue on the `pyvsita` GitHub page for this. Reading the `pyvista` source code, I see that `pyvista` extends `numpy.ndarray` and makes its own "pyvista numpy array". I need to do a refresher course on magic methods, but yes, it would be wonderful to add example docs of how to access data! Thank you for all our help!

**Comment by adamgranthendry** [comment]
@akaszynski and @MatthewFlamm I just watched [this great Intro to VTK by Kitware](https://vimeo.com/32232190) that is completely answering my question! I wished I had watched this first before reading the VTK books! Also, I really love this [Carnegie Melon Course on Discrete Differential Geometry by Prof. Keenan Crane](https://www.youtube.com/playlist?list=PL9_jI1bdZmz0hIrNCMQW1YmZysAiIYSSS). I think these resources could helps us create an intro for how to work with data types in `pyvista`.

I know I'm late to the party seeing this example (e.g. [here](https://danstoj.pythonanywhere.com/article/vtk-1#_data_in_vtk) and [here](https://materials.prace-ri.eu/391/2/VTKTutorial.pdf)). Your docs might already have this information, but if not this would be a perfect example for demoing data types! Below is a work in progress (I'll keep adding to it). Please let me know what you think:

---

## Data Types

**NOTE:** *`pyvista` uses the same data types as `VTK`, but structures them in a more pythonic manner for ease of use. For a great introduction to `VTK` data types, please see [this great Intro to VTK in Python by Kitware](https://vimeo.com/32232190), as well as the numerous code examples on [Kitware's GitHub site](https://kitware.github.io/vtk-examples/site/). An excellent introduction to mathematical concept implemented in `VTK` is provided by the [Discrete Differential Geometry YouTube Series](https://www.youtube.com/playlist?list=PL9_jI1bdZmz0hIrNCMQW1YmZysAiIYSSS) by Prof. Keenan Crane at Carnegie Melon.*

### Introduction

`VTK` is a library for visualizing (big) data. It uses an open-ended architecture of individual modules (objects) that users link together into pipelines. Data flows through a pipeline and is modified along its journey until it is displayed to screen or written to disk.

The key modules in `VTK` are:

1. **Data Structures:** *How `VTK` stores data*;
1. **Algorithms** *How `VTK` reads data from file, processes it through filters, and writes (maps) data to file or to the screen*;
1. **Renderers:** *How `VTK` uses lights, cameras, actors (data in scenes), props, and windows to display data; and* 
1. **Interactors:** *How `VTK` handles events and user inputs and wraps the output to GUI interfaces, like Qt.*


### Data Structures

![image](https://user-images.githubusercontent.com/77850046/128108464-4b2d6434-432c-4f00-a821-fd97e5356295.png)

`vtkDataObject`  &#8592;  `vtkDataSet` 

At the top-level, `vtkDataObject`s are "blob"s of data without **geometry** (*how the values are positioned in space*) or **topology** (*how the values are connected to one another*). These contain `vtkFieldData` instances (essentially scalar values) and `vtkFieldArray`s (arrays of `vtkFieldData`).

Under this are `vtkDataSet`s, which add **geometry** and **topology** to data object field data. Associated with every **point** and **cell** in the dataset is a specific value. Since these values must be positioned and connected in space, they are held in `vtkDataArrays`, which are simply memory buffers (locations on RAM where data is temporarily stored as it is being moved and processed). Technically speaking, since `VTK` is written in C++, `vtkDataArray`s live on the heap.

![image](https://user-images.githubusercontent.com/77850046/128110221-281abd42-a46b-483b-8f39-86790d9c5373.png)

The most important dataset subclasses are listed above in order from most to least complex. Every dataset can be represented as a `vtkUnstructuredGrid`, but it takes the most amount of memory. Since `vtkImageData` is uniformly spaced, it is the simplest and takes the least amount of memory to store.

![image](https://user-images.githubusercontent.com/77850046/128110868-706ac53e-f4f3-44da-93d4-83d6a2b2aaa0.png)

Though the heap is linear, `vtkDataArray`s are represented as 2D objects. Their type is determined by the data array subclass (e.g. `vtkFloatArray`, vtkIntArray`, etc.) and each consists of an array of *n* tuples by *m* components. Conceptually, you can think of them as

```
vtkDataArray = type[tuples][components]
```

### Example

As a simple example, let's pip install vtk and explore some data types:

```bash
$ py -m pip install vtk
```

We start by importing `vtk` in a REPL (IDLE, ipython, terminal, etc.) and explore the help.

```python
>>> import vtk
>>> help(vtk.vtkDoubleArray)  # Shows manpage for `vtk.vtkDoubleArray`
>>> def member_list(vtkObj):
...    """Simple helper to show methods and attributes belonging to a vtkObj.""
...    [print(member) for member in dir(vtkObj)]
>>> member_list(vtk.vtkDoubleArray)
AbstractArray
AddObserver
Allocate
AoSDataArrayTemplate
BreakOnError
COMPONENT_RANGE
ClearLookup
CopyComponent
CopyComponentNames
...
__sizeof__
__str__
__subclasshook__
__this__
__vtkname__
```

Let's make a `vtkDoubleArray()` and fill it with data:

```python
>>> # Make array
>>> dbl_arr = vtk.vtkDoubleArray()
>>> dbl_arr.SetName("my array")
>>> dbl_arr.GetName()
'my array'
>>> dbl_arr.SetNumberOfComponents(1)
>>> dbl_arr.SetNumberOfTuples(500*500)  # e.g., a 500x500 picture
>>>
>>> # Fill with data
>>> from math import sin, cos
>>> dbl_arr.SetValue(24, 5.0)  # Set the 24th tuple to 5.0
>>> dbl_arr.GetValue(24)
5.0
>>> dbl_arr.SetValue(24, 3.0)
>>> dbl_arr.GetValue(24)
3.0
>>> _ = [dbl_arr.SetValue(x*500 + y, 127.5 + (1.0 + sin(x/25.0)*cos(y/25.0))) for y in range(500) for x in range(500)]  # Assign to nothing to suppress output
>>>
>>> dbl_arr.GetValue(100)
128.5
>>> dbl_arr.GetValue(20100)
127.84663509078827
>>> print(dbl_arr)
vtkDoubleArray (0000023932CA0630)
  Debug: Off
  Modified Time: 76
  Reference Count: 1
  Registered Events: (none)      
  Name: my array
  Data type: double
  Size: 250000
  MaxId: 249999
  NumberOfComponents: 1
  Information: 0000000000000000  
  Name: my array
  Number Of Components: 1
  Number Of Tuples: 250000
  Size: 250000
  MaxId: 249999
  LookupTable: (none)
```

Now let's put this data into a dataset. We can use image data since our dataset is uniformly spaced:

```python
>>> id = vtk.vtkImageData()  # Create the data structure
>>> id.SetOrigin(0, 0, 0)
>>> id.SetSpacing(1, 1, 1)
>>>
>>> id.SetDimensions(500, 500, 1)  # Define our topology
>>>
>>> id.GetPointData().SetScalars(dbl_arr)  # Set our geometry
0
>>>
>>> print(id)  # Inspect what we got
vtkImageData (0000023932D73E80)
  Debug: Off
  Modified Time: 126
  Reference Count: 1
  Registered Events: (none)
  Information: 0000023934B751D0
  Data Released: False
  Global Release Data: Off
  UpdateTime: 0
  Field Data:
    Debug: Off
    Modified Time: 79
    Reference Count: 1
    Registered Events: (none)
    Number Of Arrays: 0
    Number Of Components: 0
    Number Of Tuples: 0
  Number Of Points: 250000
  Number Of Cells: 249001
  Cell Data:
    Debug: Off
    Modified Time: 82
    Reference Count: 1
    Registered Events:
      Registered Observers:
        vtkObserver (0000023932E88F50)
          Event: 33
          EventName: ModifiedEvent
          Command: 0000023934B75270
          Priority: 0
          Tag: 1
    Number Of Arrays: 0
    Number Of Components: 0
    Number Of Tuples: 0
    Copy Tuple Flags: ( 1 1 1 1 1 0 1 1 1 1 1 )
    Interpolate Flags: ( 1 1 1 1 1 0 0 1 1 1 1 )
    Pass Through Flags: ( 1 1 1 1 1 1 1 1 1 1 1 )
    Scalars: (none)
    Vectors: (none)
    Normals: (none)
    TCoords: (none)
    Tensors: (none)
    GlobalIds: (none)
    PedigreeIds: (none)
    EdgeFlag: (none)
    Tangents: (none)
    RationalWeights: (none)
    HigherOrderDegrees: (none)
  Point Data:
    Debug: Off
    Modified Time: 126
    Reference Count: 1
    Registered Events:
      Registered Observers:
        vtkObserver (0000023932E89370)
          Event: 33
          EventName: ModifiedEvent
          Command: 0000023934B75270
          Priority: 0
          Tag: 1
    Number Of Arrays: 1
    Array 0 name = my array
    Number Of Components: 1
    Number Of Tuples: 250000
    Copy Tuple Flags: ( 1 1 1 1 1 0 1 1 1 1 1 )
    Interpolate Flags: ( 1 1 1 1 1 0 0 1 1 1 1 )
    Pass Through Flags: ( 1 1 1 1 1 1 1 1 1 1 1 )
    Scalars:
      Debug: Off
      Modified Time: 76
      Reference Count: 2
      Registered Events: (none)
      Name: my array
      Data type: double
      Size: 250000
      MaxId: 249999
      NumberOfComponents: 1
      Information: 0000000000000000
      Name: my array
      Number Of Components: 1
      Number Of Tuples: 250000
      Size: 250000
      MaxId: 249999
      LookupTable: (none)
    Vectors: (none)
    Normals: (none)
    TCoords: (none)
    Tensors: (none)
    GlobalIds: (none)
    PedigreeIds: (none)
    EdgeFlag: (none)
    Tangents: (none)
    RationalWeights: (none)
    HigherOrderDegrees: (none)
  Bounds:
    Xmin,Xmax: (0, 499)
    Ymin,Ymax: (0, 499)
    Zmin,Zmax: (0, 0)
  Compute Time: 127
  Spacing: (1, 1, 1)
  Origin: (0, 0, 0)
  Direction: (1, 0, 0, 0, 1, 0, 0, 0, 1)
  Dimensions: (500, 500, 1)
  Increments: (0, 0, 0)
  Extent: (0, 499, 0, 499, 0, 0)
>>>
>>> print(id.GetPointData())  # Print point data
vtkPointData (0000023932B65DE0)
  Debug: Off
  Modified Time: 126
  Reference Count: 2
  Registered Events:
    Registered Observers:
      vtkObserver (0000023932E89370)
        Event: 33
        EventName: ModifiedEvent
        Command: 0000023934B75270
        Priority: 0
        Tag: 1
  Number Of Arrays: 1
  Array 0 name = my array
  Number Of Components: 1
  Number Of Tuples: 250000
  Copy Tuple Flags: ( 1 1 1 1 1 0 1 1 1 1 1 )
  Interpolate Flags: ( 1 1 1 1 1 0 0 1 1 1 1 )
  Pass Through Flags: ( 1 1 1 1 1 1 1 1 1 1 1 )
  Scalars:
    Debug: Off
    Modified Time: 76
    Reference Count: 2
    Registered Events: (none)
    Name: my array
    Data type: double
    Size: 250000
    MaxId: 249999
    NumberOfComponents: 1
    Information: 0000000000000000
    Name: my array
    Number Of Components: 1
    Number Of Tuples: 250000
    Size: 250000
    MaxId: 249999
    LookupTable: (none)
  Vectors: (none)
  Normals: (none)
  TCoords: (none)
  Tensors: (none)
  GlobalIds: (none)
  PedigreeIds: (none)
  EdgeFlag: (none)
  Tangents: (none)
  RationalWeights: (none)
  HigherOrderDegrees: (none)
>>>
>>> arr = id.GetPointData().GetArray("my array")
>>> arr
(vtkmodules.vtkCommonCore.vtkDoubleArray)0000023935E38C40
>>> arr.GetRange()
(127.5000007693025, 129.49994172022997)
>>> arr.GetTuple(1000)
(128.5799146939692,)
```

### Renderers

![image](https://user-images.githubusercontent.com/77850046/128115979-af9a2de0-82a6-44ba-b789-d3e60153f9b5.png)

**Comment by akaszynski** [comment]
That's perfect!  I'm working on web rendering right now in #1557 and it's eating up a bit of my time, plan is to add this as well.

**Comment by adamgranthendry** [comment]
@akaszynski Yay! So glad to hear it! We could add this at the next release for sure. I still need to complete the tutorial and I think we are both slammed for time at the moment. 😄
